api/agent/agent.go

```diff
@@ -447,6 +446,17 @@ func (a *agent) checkLaunch(ctx context.Context, call *call, notifyChan chan err
 
 	mem := call.Memory + uint64(call.TmpFsSize)
 
+	// In blocking mode, we attempt resource allocation in HotPoll intervals
+	// and attempt an evict if we get CapacityFull after HotPoll expiry.
+	// In non-blocking mode, the interval is set to 0.
+	resourceTimeout := a.cfg.HotPoll
+	if a.cfg.EnableNBResourceTracker {
+		resourceTimeout = 0
+	}
+
+	resourceCtx, cancelCtx := context.WithTimeout(ctx, resourceTimeout)
```

Comment:
i think it's safe to do `ctx, cancel = ` if we do it like this:

```go
if !a.cfg.EnableNBResourceTracker {
  var cancel func()
  ctx, cancel = context.WithTimeout(ctx, a.cfg.HotPoll)
  defer cancel()
}
```

this means we don't have to worry about 2 different deadlines in this scope (for `<-ctx.Done()`, the context handles that), don't have 2 contexts in scope, etc. I think this is correct, as `runHot` won't inherit timings and the state updater ignores the deadlines (just used for logging/stats). wdyt ?

Reply:
yeah, I think we have to pick basically, two contexts and pass resource specific ctx separately to resource.go or pass a timeout argument in order to avoid handling two contexts here. A timeout arg to GetResourceToken() is good/clean too. What do you think? If we use one ctx, then it gets more tricky as we don't want to accidentally bleed this specific time out into runHot() or other function calls in there...

Reply:
right. i'm wondering if there's any shifting of the channel send / receives so that we don't have to worry about this. i think if we only enter the wait loop (and launch goroutine) if we don't immediately have resources, then we can use the same context in the scope?

eg (psuedo)

```go
func GetResourceToken(ctx, params) <-chan Token {
  ch := make(chan Token, 1)
  mu.Lock()
  defer mu.Unlock() // it'll launch goroutine if we don't have resources, no worries
  ok := isResourceAvailableLocked(params)
  if ok {
    ch <- buildToken()
    return ch
  }
  
  // NOW get a chan that is unbuffered to block and launch a goroutine and return
  ch = make(chan Token)
  go awaitResourceToken(ctx, params, ch)
  return ch
}
```

eh?

Reply:
i'll clarify, this fixes the caller if we do it a certain way... 1s

Reply:
in the caller if we do:

```go
ch := GetResourceToken(ctx, params)

select {
case tok := <-ch
  return tok
default:
  // ok, we have to wait now
}

select {
case tok := <-ch:
case <-ctx.Done():
}
```

i think it works... it makes it a little easier to reason about the get resource token contract imo (but maybe not, i'm just a dude)